### PR TITLE
Add equals and hashcode to BaseSnapshot

### PR DESCRIPTION
**Pull request by Fokko**
For the Beam runner, we emit the Snapshots after they are made. This can be done to notify downstream consumers of the data. Because the snapshots will be serialized, Beam needs to have the ability to compare them. This is done by the native Java equals function. Because we implement equals, we also need to implement the hashcode function.
```
[direct-runner-worker] WARN org.apache.beam.sdk.util.MutationDetectors - Coder of type class org.apache.beam.sdk.coders.SerializableCoder has a #structuralValue method which does not return true when the encoding of the elements is equal. Element BaseSnapshot{id=8584070983395915429, timestamp_ms=1609329789570, operation=append, summary={added-data-files=3, added-records=300, added-files-size=3072, changed-partition-count=1, total-records=300, total-data-files=3, total-delete-files=0, total-position-deletes=0, total-equality-deletes=0}, manifest-list=file:/tmp/test_batch/metadata/snap-8584070983395915429-1-84dd1557-d24b-4b8b-8511-75eac387a6c5.avro}
```
I thought it would be nice to separate small changes outside of the Iceberg PR in https://github.com/apache/iceberg/pull/1985. So we keep the changes a bit smaller and makes it easier to review.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseSnapshot.java [hunk 1]
```diff
@@ -211,6 +211,32 @@ private void cacheChanges() {
...
+  }
+
+  @Override
+  public int hashCode() {
+    int hash = 7;
+    hash = 31 * hash + (int) this.snapshotId;
+    hash = 31 * hash + (this.parentId == null ? 0 : this.parentId.hashCode());
+    hash = 31 * hash + (int) sequenceNumber;
+    hash = 31 * hash + (int) timestampMillis;
+    return hash;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseSnapshot.java:237]: Ca you use `Objects.hash` instead? That's a much more compact implementation that does basically the same thing.

File: core/src/main/java/org/apache/iceberg/BaseSnapshot.java [hunk 2]
```diff
@@ -211,6 +211,32 @@ private void cacheChanges() {
...
+    }
+    if (o instanceof BaseSnapshot) {
+      BaseSnapshot other = (BaseSnapshot) o;
+      return this.snapshotId == other.snapshotId() &&
+          this.parentId.equals(other.parentId()) &&
+          this.sequenceNumber == other.sequenceNumber() &&
+          this.timestampMillis == other.timestampMillis();
+
+    }
+    return false;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseSnapshot.java:227]: Style nits: we like to separate control flow blocks with a blank line, and avoid a blank line before the end of a block or method.

File: core/src/main/java/org/apache/iceberg/BaseSnapshot.java [hunk 3]
```diff
@@ -211,6 +211,32 @@ private void cacheChanges() {
...
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o instanceof BaseSnapshot) {
+      BaseSnapshot other = (BaseSnapshot) o;
+      return this.snapshotId == other.snapshotId() &&
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseSnapshot.java:221]: Shouldn't this also include the the table  that was loaded? Or is it enough that id and parent id are almost certainly unique?

Comment (Fokko) [reply core/src/main/java/org/apache/iceberg/BaseSnapshot.java:222]: I had the same thought. For the Beam use-case, this is sufficient. If you want to split out a single stream into multiple tables, then you would do a key-by on the destination table. The comparisons will be made within the table, therefore the sequenceNumber will make this unique. I'm happy to also add the table that was loaded, but this isn't in the class right now.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseSnapshot.java:222]: I think it's okay for now, but it is an open question what makes two snapshots equal. Being careful here and including more fields as you do is pretty reasonable.

**Comment by rdblue** [comment]
The Javadoc failure was fixed in #2018 so I'm going to merge this.

**Pull request merged**